##PART 1

- You're the data scientist for a BPO company where employees use a portal for taking calls and ticketing. You are given the file 'user_id' consisting of the employee ids.
<br>
- The company is planning to change the layout and has a proposed New Version
<br>
- Currently, there are 1000 employees.

##Question 1: Your boss says that it takes money to deploy the new layout and asked you how many employees do you need to test the new layout? You found out the the effect size for the same concept in the pilot study was 0.71. What would you tell him?

In [ ]:
# Power Analysis - Determine Sample Size
import statsmodels.stats.power as smp

# Set parameters for the power analysis
effect_size = 0.71  # Specify the effect size
alpha = 0.05       # Set the significance level
power = 0.8        # Set the desired power usually set at 0.8 or 0.9 Is influenced by eff size, sample, and alpha

# Conduct power analysis for a two-sample t-test(compare means of 2 independent groups-old system/new system)
analysis = smp.TTestIndPower()
required_sample_size = analysis.solve_power(effect_size=effect_size, alpha=alpha, power=power)

# Print the result
print(f"Required sample size: {round(required_sample_size)} participants")

# Experimental Design using Power Analysis to determine sample size needed to detect an effect of a specified size with given power
# and significance level (crucial in experimental design to ensure test has high probability of detecting a true effect if there is one)

Required sample size: 32 participants


I would inform the boss that I need a sample size of 32 employees to effectively test the new layout and get statistically reliable results.

#Question 2: The boss then asked which employees would you need specifically so that the data team can pull their data for you? He needs a file consisting their user ids where you would also specify which users would have the old version and which users would they deploy the new versions to.

In [ ]:
# Experimental Design (Randomization): Randomly assigns participants to control and experimental groups for reducing bias and ensuring groups are comparable.
#print the list, export the list into a csv.


In [ ]:
import pandas as pd
df = pd.read_csv('user_id.csv')
df

user_id
0          0
1          1
2          2
3          3
4          4
..       ...
995      995
996      996
997      997
998      998
999      999

[1000 rows x 1 columns]

In [ ]:
import pandas as pd
import numpy as np

# Read the CSV file containing the employee data
data = pd.read_csv("user_id.csv")

# Display the first few rows of the DataFrame to understand its structure
print(data.head())

# Set random seed for reproducibility
np.random.seed(42)

# Define the desired sample size
sample_size = 32  # From previous analysis

# Randomly select sample_size number of employees (32)
selected_employee_ids = np.random.choice(data['user_id'], size=sample_size, replace=False)

# Assign half of the selected employees to the control group (old version) [ // is integer division ]
control_group_size = sample_size // 2
control_group_ids = selected_employee_ids[:control_group_size]

# Assign the remaining employees to the experimental group (new version)
experimental_group_ids = selected_employee_ids[control_group_size:]

# Create a DataFrame to store the user IDs and their assigned groups
employee_groups = pd.DataFrame({'user_id': selected_employee_ids,
                                'group': ['Old Version'] * control_group_size + ['New Version'] * (sample_size - control_group_size)})

# Display the DataFrame
print(employee_groups)

# Save the DataFrame to a CSV file
employee_groups.to_csv("employee_groups.csv", index=False)

print("File 'employee_groups.csv' saved successfully.")


   user_id
0        0
1        1
2        2
3        3
4        4
    user_id        group
0       521  Old Version
1       737  Old Version
2       740  Old Version
3       660  Old Version
4       411  Old Version
5       678  Old Version
6       626  Old Version
7       513  Old Version
8       859  Old Version
9       136  Old Version
10      811  Old Version
11       76  Old Version
12      636  Old Version
13      973  Old Version
14      938  Old Version
15      899  Old Version
16      280  New Version
17      883  New Version
18      761  New Version
19      319  New Version
20      549  New Version
21      174  New Version
22      371  New Version
23      527  New Version
24      210  New Version
25      235  New Version
26      101  New Version
27      986  New Version
28      902  New Version
29      947  New Version
30      346  New Version
31      139  New Version
File 'employee_groups.csv' saved successfully.


#PART 2 (FOR HOMEWORK/OPTIONAL)

use lab15_data.csv and perform all of the experiment steps you think you would need.



In [ ]:
import pandas as pd
df = pd.read_csv('lab15_data.csv')
df

user_id version  new_user_activation  purchases
0          0       A                False          1
1          1       A                 True          1
2          2       A                 True          3
3          3       A                False          0
4          4       A                 True          0
..       ...     ...                  ...        ...
995      995       A                False          0
996      996       B                 True          0
997      997       A                False          3
998      998       B                False          1
999      999       B                 True          0

[1000 rows x 4 columns]

In [ ]:
# Randomly Select Required Sample size

# Set random seed for reproducibility
np.random.seed(42)

# Define the desired sample size
sample_size = 32  # From previous analysis

# Randomly select sample_size number of employees
selected_employee_ids = np.random.choice(data['user_id'], size=sample_size, replace=False)

# Create a DataFrame with the selected employee IDs
selected_employees = data[data['user_id'].isin(selected_employee_ids)]

# Display the selected employees
print(selected_employees)

     user_id version  new_user_activation  purchases
76        76       B                 True          1
101      101       B                 True          2
136      136       A                 True          3
139      139       A                 True          0
174      174       A                 True          1
210      210       B                False          1
235      235       A                False          0
280      280       B                 True          0
319      319       B                False          1
346      346       A                 True          0
371      371       B                 True          0
411      411       B                 True          1
513      513       B                 True          1
521      521       A                False          3
527      527       B                 True          0
549      549       A                 True          1
626      626       B                False          0
636      636       B                False     

In [ ]:
# Randomly Assign Sample to Control Experimental Groups

# Determine the size of the control group
control_group_size = sample_size // 2

# Randomly assign half of the selected employees to the control group
control_group_ids = selected_employee_ids[:control_group_size]

# Assign the remaining employees to the experimental group
experimental_group_ids = selected_employee_ids[control_group_size:]

# Create a new column 'group' and initialize it with 'Old Version' for control group
selected_employees.loc[selected_employees['user_id'].isin(control_group_ids), 'group'] = 'Old Version'

# Update the 'group' column to 'New Version' for experimental group
selected_employees.loc[selected_employees['user_id'].isin(experimental_group_ids), 'group'] = 'New Version'

# Display the selected employees with their assigned groups
print(selected_employees)


     user_id version  new_user_activation  purchases        group
76        76       B                 True          1  Old Version
101      101       B                 True          2  New Version
136      136       A                 True          3  Old Version
139      139       A                 True          0  New Version
174      174       A                 True          1  New Version
210      210       B                False          1  New Version
235      235       A                False          0  New Version
280      280       B                 True          0  New Version
319      319       B                False          1  New Version
346      346       A                 True          0  New Version
371      371       B                 True          0  New Version
411      411       B                 True          1  Old Version
513      513       B                 True          1  Old Version
521      521       A                False          3  Old Version
527      5

<ipython-input-18-9df03958c2c0>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_employees.loc[selected_employees['user_id'].isin(control_group_ids), 'group'] = 'Old Version'


In [ ]:
# Simulate Performance Data A/B testing: Simulate Performance Data

np.random.seed(42)
performance_old_version = np.random.normal(loc=70, scale=10, size=control_group_size)
performance_new_version = np.random.normal(loc=75, scale=10, size=sample_size - control_group_size)

# Assign the performance data to the DataFrame
selected_employees.loc[selected_employees['group'] == 'Old Version', 'performance_score'] = performance_old_version
selected_employees.loc[selected_employees['group'] == 'New Version', 'performance_score'] = performance_new_version

# Display the selected employees with their performance scores
print(selected_employees)


     user_id version  new_user_activation  purchases        group  \
76        76       B                 True          1  Old Version   
101      101       B                 True          2  New Version   
136      136       A                 True          3  Old Version   
139      139       A                 True          0  New Version   
174      174       A                 True          1  New Version   
210      210       B                False          1  New Version   
235      235       A                False          0  New Version   
280      280       B                 True          0  New Version   
319      319       B                False          1  New Version   
346      346       A                 True          0  New Version   
371      371       B                 True          0  New Version   
411      411       B                 True          1  Old Version   
513      513       B                 True          1  Old Version   
521      521       A              

<ipython-input-19-c6684a9ddbd9>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_employees.loc[selected_employees['group'] == 'Old Version', 'performance_score'] = performance_old_version


In [ ]:
# Analyze Results of A/B Test

from scipy.stats import ttest_ind

# Separate the performance scores into two groups
performance_old_version = selected_employees[selected_employees['group'] == 'Old Version']['performance_score']
performance_new_version = selected_employees[selected_employees['group'] == 'New Version']['performance_score']

# Perform a two-sample t-test
t_stat, p_value = ttest_ind(performance_old_version, performance_new_version)

# Print the results of the t-test
print(f"T-statistic: {t_stat}")
print(f"P-value: {p_value}")

# Interpret the results
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in performance between the old and new layouts.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in performance between the old and new layouts.")


T-statistic: -0.8185105831753815
P-value: 0.4195201489501176
Fail to reject the null hypothesis: There is no significant difference in performance between the old and new layouts.


P-value 0.42 > significance(alpha, 0.05) we fail to reject the null - there is no statistically significant difference in performance between the employees using old and new layouts.


*Note: this is a different dataset*